src/node/setupRemoteServer.ts

```diff
@@ -0,0 +1,212 @@
+import * as http from 'http'
+import { invariant } from 'outvariant'
+import { Server as WebSocketServer } from 'socket.io'
+import type { Socket } from 'socket.io-client'
+import { Emitter } from 'strict-event-emitter'
+import { DeferredPromise } from '@open-draft/deferred-promise'
+import {
+  LifeCycleEventsMap,
+  RequestHandler,
+  SetupApi,
+  handleRequest,
+  rest,
+} from '~/core'
+import {
+  SerializedRequest,
+  SerializedResponse,
+  deserializeRequest,
+  deserializeResponse,
+  serializeRequest,
+  serializeResponse,
+} from '~/core/utils/request/serializeUtils'
+import { LifeCycleEventEmitter } from '~/core/sharedOptions'
+
+export const SYNC_SERVER_URL = new URL('http://localhost:50222')
+
+export function setupRemoteServer(...handlers: Array<RequestHandler>) {
+  return new SetupRemoteServerApi(...handlers)
+}
+
+export interface SetupRemoteServer {
+  listen(): Promise<void>
+  close(): Promise<void>
+
+  events: LifeCycleEventEmitter<LifeCycleEventsMap>
+}
+
+export interface SyncServerEventsMap {
+  request(
+    serializedRequest: SerializedRequest,
+    requestId: string,
+  ): Promise<void> | void
+  response(serializedResponse?: SerializedResponse): Promise<void> | void
+}
+
+declare global {
+  var syncServer: WebSocketServer<SyncServerEventsMap> | undefined
+}
+
+export class SetupRemoteServerApi
+  extends SetupApi<LifeCycleEventsMap>
+  implements SetupRemoteServer
+{
+  protected handlers: Array<RequestHandler>
+  protected emitter: Emitter<LifeCycleEventsMap>
+
+  constructor(...handlers: Array<RequestHandler>) {
+    super(...handlers)
+
+    this.handlers = handlers
+    this.emitter = new Emitter()
+  }
+
+  public async listen(): Promise<void> {
+    const server = await createSyncServer()
+
+    server.on('connection', (socket) => {
+      socket.on('request', async (serializedRequest, requestId) => {
+        const request = deserializeRequest(serializedRequest)
+        const response = await handleRequest(
+          request,
+          requestId,
+          this.handlers,
+          { onUnhandledRequest() {} },
+          this.emitter,
+        )
+
+        socket.emit(
+          'response',
+          response ? await serializeResponse(response) : undefined,
+        )
+      })
+
+      /**
+       * @todo Have the socket signal back whichever response
+       * was used for whichever request. Include request ID
+       * and somehow let this API know whether the response was
+       * the mocked one or note.
+       */
+      // socket.on('response', (serializedResponse) => {
```

Comment:
The client socket must send some read-only events indicating that the response has occurred.